### PR TITLE
Bug 1075420 - 'git not found' when validate scalable app with database via oo-idler-stats.

### PR DESCRIPTION
**Pull request by dinhxuanvu**
Scalable app with database has each cartridge on separate gear and some gears with database
cartridge should not contain git repo. When oo-idler-stats checks those gears, it will fail
to find git repo.

This commit modifies oo-idler-stats to only look for git repo for web_framework cartridge
and ignores git check for other types of cartridges.

This commit also fixes various broken codes such as cartridge_types method does not work
at all and run method also returns a list with empty string due to mis-use split method.

Bug 1075420
Link https://bugzilla.redhat.com/show_bug.cgi?id=1075420

Signed-off-by: Vu Dinh vdinh@redhat.com

### CODE REVIEW

**Comment by dinhxuanvu** [comment]
@tiwillia @Miciah Please review and [test]. Thanks!

File: node-util/sbin/oo-idler-stats [hunk 1]
```diff
@@ -131,7 +130,7 @@ def percent(a, b):
 
 
 def cartridge_types(homedir):
-    return_code, types = run("ls |egrep -v '-|git'", homedir)
+    return_code, types = run("ls |egrep -v -e '-|git'", homedir)
```
Comment (tiwillia) [inline node-util/sbin/oo-idler-stats]: Could you explain why this change was necessary? The previous syntax looks correct according to the egrep manpage and some quick testing.

Comment (dinhxuanvu) [inline node-util/sbin/oo-idler-stats]: @tiwillia That code didn't work as it failed while I was testing it. Apparently, due to the pattern starting with '-' in '-|git', it's recommended to use -e to avoid the egrep misinterpret the pattern as a option. It's mentioned on manpage.

Comment (tiwillia) [inline node-util/sbin/oo-idler-stats]: Ah, that does make sense. Thank you.

**Comment by tiwillia** [comment]
@Miciah or @dobbymoodge could one of y'all review? This lgtm, but I'm not extremely familiar with python and do not feel like I would be able to catch most potential issues.

File: node-util/sbin/oo-idler-stats [hunk 2]
```diff
@@ -144,40 +143,65 @@ def get_app_type(app_path):
...
+        # Check if cartridge has been checked for cartridge type before
+        if cart in cartridges_list:
+            # Flag git_repo if this cartridge should have a git repo
+            if cartridge_types.get(cart) == true:
+                git_repo = True
+                break
+        else:
+            cmd = " ".join(["cat",os.path.join(app_path, cart, "metadata/manifest.yml"),
+                                               "| grep web_framework"])
+            ignore, result = run(cmd)
```
Comment (dobbymoodge) [inline node-util/sbin/oo-idler-stats]: It's more "pythonic" to use "_" instead of "ignore" for placeholders when unpacking tuples.

**Comment by dinhxuanvu** [comment]
@jwhonce I would like to have your inputs on this PR since it involves your original codes. Thanks!

**Comment by dinhxuanvu** [comment]
@jwhonce I have removed the optimization part to force the method to check yaml files in all cases to avoid the issue with potential false reports on different same-name cartridge. Let me know if you find any other issues. Thanks.

**Comment by dinhxuanvu** [comment]
re-[test]

**Comment by dinhxuanvu** [comment]
re-[test] please.

**Comment by dinhxuanvu** [comment]
@jwhonce I wonder if I have your blessing to merge this?

**Comment by jwhonce** [comment]
LGTM

**Comment by dinhxuanvu** [comment]
@Miciah @tiwillia I have blessing from Jhon, would you mind merging this? Thanks!

File: node-util/sbin/oo-idler-stats [hunk 3]
```diff
@@ -144,40 +143,54 @@ def get_app_type(app_path):
...
-
-    # Is there stuff inside git/appname.git?
-    gitfiles = os.listdir(app_path + "/git/" + app_name + ".git")
-    if len(gitfiles) == 0:
-        return "Path %s/git/%s.git is empty." % (app_path, app_name)
-
-    # Chances are good that this is a git repo. What's the last commit?
-    last_commit = run('git show HEAD | grep Date | sed "s/Date:   //"',
-                      app_path + "/git/" + app_name + ".git")[1][0]
+    global app_name
```
Comment (Miciah) [inline node-util/sbin/oo-idler-stats]: Why make `app_name` a global?

Comment (dinhxuanvu) [inline node-util/sbin/oo-idler-stats]: app_name is a variable out of the scope of the method. In order to modify it inside the method, it needs to be global. This is for reusing the app name that is collected from the git repo from the first gear for the subsequent gears from the same app which may not have git repo.

Comment (Miciah) [inline node-util/sbin/oo-idler-stats]: This means that you are assuming that `validate` will be called first on a gear with a framework cartridge, and then subsequently on any other gears from the same application, correct? Looking at https://github.com/dinhxuanvu/origin-server/blob/idler-stats/node-util/sbin/oo-idler-stats#L231-L247, it is not obvious to me (1) that gears will be grouped by application and (2) that gears will be sorted such that a gear with a framework cartridge comes before related gears without.

**Comment by dinhxuanvu** [comment]
@Miciah [test] & ready for review.

**Comment by Miciah** [comment]
Please [merge]!
